### PR TITLE
ML Model Interface

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description
This will allow us to convert our remote models into a strong-type-like model, which we use an `input` and `output` in the `model_interface` field to annotate the input and the output schema for the model. Notably, this interface can also be used to validate the input and output of the model.
 
### Issues Resolved
This issue is for the RFC #2354 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 1]
```diff
@@ -160,8 +160,12 @@ public class CommonValue {
                         + "\" : {\"type\": \"flat_object\"},\n"
                         + "      \""
                         + AbstractConnector.ACTIONS_FIELD
+                        + "\" : {\"type\": \"flat_object\"},\n"
+                        + "      \""
+                        + AbstractConnector.MODEL_INTERFACE_FIELD
                         + "\" : {\"type\": \"flat_object\"}\n";
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:166]: From my understanding, this interface field is for model, not only for connectors but also local models like text_embeddings etc. Why do we define it only in connector, but in model?

File: common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java [hunk 2]
```diff
@@ -127,6 +129,9 @@ public HttpConnector(String protocol, XContentParser parser) throws IOException
                 case CLIENT_CONFIG_FIELD:
                     connectorClientConfig = ConnectorClientConfig.parse(parser);
                     break;
+                case MODEL_INTERFACE_FIELD:
+                    modelInterface =  getInterfaceMap(parser.map());
+                    break;
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java:134]: Just want to double confirm that we don't have to version control for it, right?

File: common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java [hunk 3]
```diff
@@ -304,6 +321,9 @@ public void update(MLCreateConnectorInput updateContent, Function<String, String
         if (updateContent.getConnectorClientConfig() != null) {
             this.connectorClientConfig = updateContent.getConnectorClientConfig();
         }
+        if (updateContent.getModelInterface() != null && updateContent.getModelInterface().size() > 0) {
+            this.modelInterface = updateContent.getModelInterface();
+        }
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java:326]: Why don't we allow updating interface to empty map?

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 4]
```diff
@@ -95,6 +96,28 @@ public static Map<String, Object> fromJson(String jsonStr, String defaultKey) {
         return result;
     }
 
+    public static Map<String, String> getInterfaceMap(Map<String, ?> interfaceObjs) {
+        Map<String, String> parameters = new HashMap<>();
+        for (String key : interfaceObjs.keySet()) {
+            if (Objects.equals(key, "input") || Objects.equals(key, "output")) {
```
Comment (jngz-es) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:102]: Recommend using stream.filter instead of for loop.

**Comment by jngz-es** [comment]
@b4sjoo , could you share an example for input/output config? That will help us to understand the pr better.

File: ml-algorithms/build.gradle [hunk 5]
```diff
@@ -28,6 +28,7 @@ dependencies {
        exclude group: "org.jetbrains", module: "annotations"
     }
     implementation group: 'org.apache.commons', name: 'commons-text', version: '1.10.0'
+    implementation group: 'com.github.erosb', name: 'json-sKema', version: '0.15.0'
```
Comment (jngz-es) [inline ml-algorithms/build.gradle:31]: Why don't we use https://github.com/json-schema-org/json-schema-spec?
The package "com.github.erosb.json-sKema" looks like having very few references, not sure if it is prod ready.

Comment (b4sjoo) [reply ml-algorithms/build.gradle:30]: Changed to top used library.

File: common/src/main/java/org/opensearch/ml/common/connector/AbstractConnector.java [hunk 6]
```diff
@@ -44,6 +44,7 @@ public abstract class AbstractConnector implements Connector {
     public static final String OWNER_FIELD = "owner";
     public static final String ACCESS_FIELD = "access";
     public static final String CLIENT_CONFIG_FIELD = "client_config";
+    public static final String MODEL_INTERFACE_FIELD = "model_interface";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/AbstractConnector.java:47]: Move to model layer and change field name to "interface"?

File: common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java [hunk 7]
```diff
@@ -269,6 +280,12 @@ public void writeTo(StreamOutput out) throws IOException {
         } else {
             out.writeBoolean(false);
         }
+        if (modelInterface != null) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java:283]: Add BWC version?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java [hunk 8]
```diff
@@ -252,4 +263,37 @@ public static SdkHttpFullRequest signRequest(
...
+
+    public static String parametersObjectExtractor(MLInput mlInput) throws IOException {
+        String mlInputString = mlInput.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS).toString();
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode mlInputObject = mapper.readTree(mlInputString);
+        return mlInputObject.get("parameters").asText();
+    }
+
+    public static void validateSchema(String schemaString, String instanceString) throws IOException {
+        System.out.println(schemaString);
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:275]: Remove this line, use log if need to print log

File: common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java (hunk 7 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java:283]: Not here

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java (hunk 8 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java:275]: Sure

File: common/src/main/java/org/opensearch/ml/common/connector/AbstractConnector.java (hunk 6 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/connector/AbstractConnector.java:47]: Ack

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java (hunk 1 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/CommonValue.java]: Moved out to Model Index

File: build.gradle [hunk 9]
```diff
@@ -23,7 +23,7 @@ buildscript {
         }
 
         common_utils_version = System.getProperty("common_utils.version", opensearch_build)
-        kotlin_version = System.getProperty("kotlin.version", "1.8.21")
+        kotlin_version = System.getProperty("kotlin.version", "1.9.23")
```
Comment (ylwu-amzn) [inline build.gradle:26]: @Hailong-am added Kotlin to ml-commons in this PR https://github.com/opensearch-project/ml-commons/pull/1845 for search alert tool

Hailong, do we really need this dependency now?

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 10]
```diff
@@ -265,45 +265,48 @@ public class CommonValue {
...
-                "                    \"source_fields\": {\n" +
-                "                      \"type\": \"text\"\n" +
-                "                    }\n" +
-                "                  }\n" +
-                "                }\n" +
-                "              }\n" +
-                "            }\n" +
-                "          }\n" +
-                "        },\n"
+                        "          \"properties\": {\n" +
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:273]: not a big problem , but better to make the code consistent. The lines above this line has `+` in the beginning of line, can we keep it consistent? It's fine to fix in a new PR if no enough bandwidth.

File: common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java [hunk 11]
```diff
@@ -117,6 +118,20 @@ public void getParameterMap() {
         assertEquals("[1.01,\"abc\"]", parameterMap.get("key5"));
     }
 
+    @Test
+    public void getInterfaceMap() {
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("input", "value1");
+        parameters.put("output", 2);
+        parameters.put("key3", 2.1);
```
Comment (mingshl) [inline common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:126]: so the keys which are not "input" and "output" will be ignored? users can define a model interface but not using "input" and "output" then it would be empty model interface? is it the intention here?

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 12]
```diff
@@ -89,6 +93,7 @@ public class MLModel implements ToXContentObject {
     public static final String CONNECTOR_FIELD = "connector";
     public static final String CONNECTOR_ID_FIELD = "connector_id";
     public static final String GUARDRAILS_FIELD = "guardrails";
+    public static final String MODEL_INTERFACE_FIELD = "model_interface";
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:96]:
```suggestion
    public static final String MODEL_INTERFACE_FIELD = "interface";
```

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 13]
```diff
@@ -261,6 +270,9 @@ public MLModel(StreamInput input) throws IOException {
             if (input.readBoolean()) {
                 this.guardrails = new Guardrails(input);
             }
+            if (input.readBoolean()) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:273]: Add BWC version?

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java [hunk 14]
```diff
@@ -116,51 +124,16 @@ public MLUpdateModelInput(StreamInput in) throws IOException {
...
+            if (in.readBoolean()) {
+                modelInterface = in.readMap(StreamInput::readString, StreamInput::readString);
+            }
+        }
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        builder.field(MODEL_ID_FIELD, modelId);
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java:124]: Why remove these lines?

File: build.gradle (hunk 9 above)
Comment (Hailong-am) [reply build.gradle:26]: yes, we still need it. You can see https://github.com/opensearch-project/skills/issues/116 for more context. As long as ml-commons has dependency on common-utils, kotlin dependency is needed for skills repo.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 15]
```diff
@@ -95,6 +103,20 @@ public static Map<String, Object> fromJson(String jsonStr, String defaultKey) {
         return result;
     }
 
+    public static Map<String, String> filterInterfaceMap(Map<String, ?> interfaceObjs) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:106]: Why put this model interface specific method in this generic StringUtils class ?

Or why build this function specially for model interface ? If you want to put it in this generic StringUtils class, please make this function generic too. Otherwise you can move to MLModel class

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/AwsConnectorExecutor.java [hunk 16]
```diff
@@ -68,6 +68,9 @@ public class AwsConnectorExecutor extends AbstractConnectorExecutor {
     @Setter
     @Getter
     private MLGuard mlGuard;
+    @Setter
+    @Getter
+    private Map<String, String> modelInterface;
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/AwsConnectorExecutor.java:73]: Curious why put this in connector executor while interface is on model level?

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java [hunk 17]
```diff
@@ -183,10 +183,44 @@ public TokenBucket getUserRateLimiter(String modelId, String user) {
...
     }
 
+    /**
+     * Set the ml interface for the model
+     *
+     * @param modelId model id
+     * @param modelInterface model interface
+     */
+    public synchronized void setModelInterface(String modelId, Map<String, String> modelInterface) {
+        log.debug("Setting ML Interface {} for Model {}", modelInterface, modelId);
```
Comment (mingshl) [inline plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java:193]: wondering if there is IT test for connector and check for the response to get connector, and see if the model interface is in the response body?

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 12 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:96]: changed

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 13 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:273]: No need here

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 15 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:106]: Re-write to a more generic method

File: common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java (hunk 11 above)
Comment (b4sjoo) [reply common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:126]: Yes you are right

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java (hunk 17 above)
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelCacheHelper.java:193]: Per Yaliang's request we are going to move the check to rest layer

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 18]
```diff
@@ -160,7 +160,7 @@ public class CommonValue {
                         + "\" : {\"type\": \"flat_object\"},\n"
                         + "      \""
                         + AbstractConnector.ACTIONS_FIELD
-                        + "\" : {\"type\": \"flat_object\"}\n";
+                        + "\" : {\"type\": \"flat_object\"},\n";
```
Comment (zane-neo) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:163]: Adding comma here makes the `ML_CONNECTOR_INDEX_FIELDS` isn't a json, did we test this change?

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java (hunk 10 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/CommonValue.java:273]: Looks like this part has different indention, please run `./gradlew spotlessApply` to format code. Also, I suggest we can move these configurations to a resource file with json suffix, it would be much easier to read and maintain.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 12 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:96]: interface is a general/high level word which can be misleading, a suggestion would be `api_specs` which can clearly imply this field is defining `specifications`.

File: common/src/main/java/org/opensearch/ml/common/transport/register/MLRegisterModelInput.java [hunk 19]
```diff
@@ -372,6 +396,7 @@ public static MLRegisterModelInput parse(XContentParser parser, String modelName
         Boolean doesVersionCreateModelGroup = null;
         Boolean isHidden = null;
         Guardrails guardrails = null;
+        Map<String, String> modelInterface = new HashMap<>();
```
Comment (zane-neo) [inline common/src/main/java/org/opensearch/ml/common/transport/register/MLRegisterModelInput.java:399]: Init it to null to avoid meaningless object creation since `filterInterfaceMap` will always return an object.

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 15 above)
Comment (zane-neo) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:106]: Just use `map.get("input/output")` and you can save object creation and extra for loops.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 20]
```diff
@@ -134,6 +139,8 @@ public class MLModel implements ToXContentObject {
     private String connectorId;
     private Guardrails guardrails;
 
+    private Map<String, String> modelInterface;
```
[the remark this answers is not in the document]
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:146]: Yes

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java (hunk 18 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/CommonValue.java]: Fixed.

File: common/src/main/java/org/opensearch/ml/common/transport/register/MLRegisterModelInput.java (hunk 19 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/transport/register/MLRegisterModelInput.java:400]: Changed

File: ml-algorithms/build.gradle [hunk 21]
```diff
@@ -24,9 +24,7 @@ dependencies {
     implementation "org.opensearch.client:opensearch-rest-client:${opensearch_version}"
     testImplementation "org.opensearch.test:framework:${opensearch_version}"
     implementation "org.opensearch:common-utils:${common_utils_version}"
-    implementation ("org.jetbrains.kotlin:kotlin-stdlib:${kotlin_version}") {
-       exclude group: "org.jetbrains", module: "annotations"
-    }
+    implementation ("org.jetbrains.kotlin:kotlin-stdlib:${kotlin_version}")
```
Comment (Hailong-am) [inline ml-algorithms/build.gradle:27]: is this change by intention?

File: plugin/build.gradle [hunk 22]
```diff
@@ -335,6 +336,10 @@ configurations.all {
     resolutionStrategy.force 'commons-codec:commons-codec:1.15'
     resolutionStrategy.force 'org.slf4j:slf4j-api:1.7.36'
     resolutionStrategy.force 'org.codehaus.plexus:plexus-utils:3.3.0'
+    resolutionStrategy.force 'com.fasterxml.jackson.core:jackson-databind:2.17.0'
+    resolutionStrategy.force 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.17.0'
```
Comment (Hailong-am) [inline plugin/build.gradle:340]: can the hardcode 2.17.0 be replaced by `${versions.jackson}`?

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java (hunk 14 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java:124]: Add java doc to this method and add this comment. It's easy to forget why

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 23]
```diff
@@ -134,6 +143,8 @@ public class MLModel implements ToXContentObject {
...
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:146]: Add some example ?

File: common/src/test/java/org/opensearch/ml/common/transport/model/MLUpdateModelInputTest.java [hunk 24]
```diff
@@ -85,7 +84,7 @@ public class MLUpdateModelInputTest {
             "{\"model_type\":\"testModelType\",\"embedding_dimension\":100,\"framework_type\":\"SENTENCE_TRANSFORMERS\",\"all_config\":\""
             +
             "{\\\"field1\\\":\\\"value1\\\",\\\"field2\\\":\\\"value2\\\"}\"},\"connector_id\":" +
-            "\"test-connector_id\",\"connector\":{\"description\":\"updated description\",\"version\":\"1\",\"parameters\":{},\"credential\":{}}}";
+            "\"test-connector_id\",\"interface\":{}}";
```
Comment (ylwu-amzn) [inline common/src/test/java/org/opensearch/ml/common/transport/model/MLUpdateModelInputTest.java:87]: Add some test case for a non-empty interface ?

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 25]
```diff
@@ -95,6 +105,28 @@ public static Map<String, Object> fromJson(String jsonStr, String defaultKey) {
         return result;
     }
 
+    public static Map<String, String> filteredParameterMap(Map<String, ?> parameterObjs, Set<String> allowedList) {
+        Map<String, String> parameters = new HashMap<>();
+        Set<String> filteredKeys = new HashSet<>(parameterObjs.keySet());
+        filteredKeys.retainAll(allowedList);
+        for (String key : filteredKeys) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:112]: Seems no need to add a new HashSet `filteredKeys`, just need to iterate the `allowedList` 
```
    for (String key : allowedList) {
        if (parameterObjs.containsKey(key)) {
```

File: common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java [hunk 26]
```diff
@@ -117,6 +120,21 @@ public void getParameterMap() {
...
+        final Set<String> allowedInterfaceFieldNameList = new HashSet<>(Arrays.asList("input","output"));
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("input", "value1");
+        parameters.put("output", 2);
+        parameters.put("key3", 2.1);
+        parameters.put("key4", new int[]{10, 20});
+        parameters.put("key5", new Object[]{1.01, "abc"});
+        Map<String, String> interfaceMap = StringUtils.filteredParameterMap(parameters, allowedInterfaceFieldNameList);
+        Assert.assertEquals(2, interfaceMap.size());
+        Assert.assertEquals("value1", interfaceMap.get("input"));
```
Comment (ylwu-amzn) [inline common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:134]: This use case is too simple. Add some real json schema use case for the input/output ?

File: ml-algorithms/build.gradle [hunk 27]
```diff
@@ -24,9 +24,7 @@ dependencies {
     implementation "org.opensearch.client:opensearch-rest-client:${opensearch_version}"
     testImplementation "org.opensearch.test:framework:${opensearch_version}"
     implementation "org.opensearch:common-utils:${common_utils_version}"
-    implementation ("org.jetbrains.kotlin:kotlin-stdlib:${kotlin_version}") {
-       exclude group: "org.jetbrains", module: "annotations"
```
Comment (ylwu-amzn) [inline ml-algorithms/build.gradle:28]: @b4sjoo Why remove this part ?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java [hunk 28]
```diff
@@ -34,23 +35,20 @@
 
 public interface RemoteConnectorExecutor {
 
-    default ModelTensorOutput executePredict(MLInput mlInput) {
+    default ModelTensorOutput executePredict(MLInput mlInput) throws IOException {
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:38]: Why add `throws IOException` ?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java [hunk 29]
```diff
@@ -150,5 +148,4 @@ && getUserRateLimiterMap().get(user.getName()) != null
     }
 
     void invokeRemoteModel(MLInput mlInput, Map<String, String> parameters, String payload, List<ModelTensors> tensorOutputs);
-
```
Comment (ylwu-amzn) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:153]: Is this required by checkstyle to remove this line ?  Why `ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java` requires a new line at line 255?

File: plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java [hunk 30]
```diff
@@ -227,4 +237,24 @@ private void executePredict(
...
+                throw new IllegalArgumentException("Error validating input schema: " + e.getMessage());
+            }
+        }
+    }
+
+    private static String parametersObjectExtractor(MLInput mlInput) throws IOException {
+        String mlInputString = mlInput.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS).toString();
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode mlInputObject = mapper.readTree(mlInputString);
+        return mlInputObject.get("parameters").asText();
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java:257]: By hard code `parameters` here, this function can only be used by remote model, and we can't extend for future use case, for example what if we add another field in model `input` like
```
{
    "parameters": {
        "input1": "..."
    },
    "new_field1": "..." // new field added 
}
```

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 25 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:112]: `filteredKeys` is the intersection of Map keys and allowedList, it will be shorter.

File: common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java (hunk 26 above)
Comment (b4sjoo) [reply common/src/test/java/org/opensearch/ml/common/utils/StringUtilsTest.java:134]: This method only filtered keys.

File: ml-algorithms/build.gradle (hunk 27 above)
Comment (b4sjoo) [reply ml-algorithms/build.gradle:28]: I moved this exclusion declaration to plugin level to exclude this package globally.

File: plugin/build.gradle (hunk 22 above)
Comment (b4sjoo) [reply plugin/build.gradle:340]: Removed.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java (hunk 29 above)
Comment (b4sjoo) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java:153]: Reverted.

File: plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java (hunk 30 above)
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java:257]: Removed

File: common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java (hunk 14 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/transport/model/MLUpdateModelInput.java:124]: Added back

**Review by ylwu-amzn — APPROVED** [review]

File: common/src/test/java/org/opensearch/ml/common/transport/model/MLUpdateModelInputTest.java (hunk 24 above)
Comment (b4sjoo) [reply common/src/test/java/org/opensearch/ml/common/transport/model/MLUpdateModelInputTest.java]: See UTs in task runner and trasport prediction

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 23 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:182]: Added

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 31]
```diff
@@ -143,6 +143,42 @@ public class MLModel implements ToXContentObject {
     private String connectorId;
     private Guardrails guardrails;
 
+    /**
+     * Model interface is a map that contains the input and output fields of the model, with JSON schema as the value.
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:147]: Can user just define input or output ?

File: common/src/main/java/org/opensearch/ml/common/MLModel.java [hunk 32]
```diff
@@ -143,6 +143,42 @@ public class MLModel implements ToXContentObject {
...
+     *           "properties": {
+     *             "messages": {
+     *               "type": "string",
+     *               "description": "This is a test description field"
+     *             }
+     *           }
+     *         }
+     *       }
+     *     },
+     *     "output": {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/MLModel.java:163]: This just defines output part of remote model , seems not enough, as you can see `"inference_results"` is array, if we just define the inner `output`, user won't know the inference_results structure 
```
{
  "inference_results": [
    {
      "output": [
        {
          "name": "response",
          "dataAsMap": {
            "completion": " There is no single, universally agreed upon meaning of life. The meaning of life is subjective and personal. Some common perspectives include finding happiness, purpose, spiritual fulfillment, connecting with others, contributing value, and leaving a positive legacy. Ultimately, the meaning of life is what you make of it.",
            "stop_reason": "stop_sequence"
          }
        }
      ]
    }
  ]
}
```

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 31 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:147]: Yes, user can define it separately, and check will start separately. No need to define all

File: common/src/main/java/org/opensearch/ml/common/CommonValue.java [hunk 33]
```diff
@@ -265,45 +262,48 @@ public class CommonValue {
                         + MLModel.LAST_UNDEPLOYED_TIME_FIELD
                         + "\": {\"type\": \"date\", \"format\": \"strict_date_time||epoch_millis\"},\n"
                         + "      \""
+                        + MLModel.INTERFACE_FIELD
+                        + "\": {\"type\": \"flat_object\"},\n"
+                        + "      \""
```
Comment (Zhangxunmt) [inline common/src/main/java/org/opensearch/ml/common/CommonValue.java:267]: I see that allowedInterfaceFieldKeys only includes "input", "output". Will new allowed keys possible added in the future?  If not, why not just define a Class rather than a Map<String, String> modelInterface? Using the Map is primarily for extensibility.

File: common/src/main/java/org/opensearch/ml/common/MLModel.java (hunk 32 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/MLModel.java:163]: Changed.

**Review by ylwu-amzn — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]
